### PR TITLE
Cleanup httpmock

### PR DESCRIPTION
**Pull request by samcoe**
This PR gets the package formerly known as `httpmock` into a state that it can be made public and moved out of the `internal` folder. 

Some major changes:
- Change package name to `transportmock`
- Require names for stubs
- Removed extraneous responders
- Allow responders to handle strings and interface input 

I left comments for open questions inline.

cc https://github.com/cli/go-gh/issues/28

### CODE REVIEW

File: internal/transportmock/matcher.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	"encoding/json"
+	"io"
+	"net/http"
+	"regexp"
+	"strings"
+)
+
+type Matcher func(req *http.Request) bool
+
+func REST(method, p string) Matcher {
```
Comment (samcoe) [inline internal/transportmock/matcher.go:14]: What do we think about the naming scheme here? Should we prefix/postfix `Match`/`Matcher` to this and the `GQL` function?

File: internal/transportmock/registry.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	"net/http"
+	"net/http/httputil"
+	"sync"
+	"testing"
+)
+
+type Registry struct {
+	mu       sync.Mutex
+	stubs    []*stub
+	requests []*http.Request
```
Comment (samcoe) [inline internal/transportmock/registry.go:14]: I purposefully un-exported these and added a `Requests()` accessor function. It felt smart to move this direction incase we want to change the underlying implementation.

File: internal/transportmock/registry.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+func (r *Registry) RoundTrip(req *http.Request) (*http.Response, error) {
+	var stub *stub
+	r.mu.Lock()
+	for _, s := range r.stubs {
+		if s.matched || !s.matcher(req) {
+			continue
+		}
+		if stub != nil {
+			r.mu.Unlock()
+			dr, _ := httputil.DumpRequestOut(req, true)
```
Comment (samcoe) [inline internal/transportmock/registry.go:57]: I think using `DumpRequestOut` leads to nicer looking error messages.

File: internal/transportmock/responder.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+)
+
+type Responder func(req *http.Request) (*http.Response, error)
+
+func HTTPResponse(status int, header *map[string][]string, body interface{}, cb func(*http.Request)) Responder {
```
Comment (samcoe) [inline internal/transportmock/responder.go:12]: What do we think about the naming scheme for Responders? Should we prefix/postfix them with `Response`/`Responder`?

File: internal/transportmock/responder.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+			cb(req)
+		}
+		if header == nil {
+			header = &map[string][]string{}
+		}
+		return httpResponse(status, *header, b, req), nil
+	}
+}
+
+func RESTResponse(body interface{}, cb func(map[string]interface{})) Responder {
```
Comment (samcoe) [inline internal/transportmock/responder.go:31]: I think having `HTTPResponse` and `RESTResponse` to be a bit confusing. Open to better naming here.

Comment (heaths) [reply internal/transportmock/responder.go:31]: Any less confusing if you call `RESTResponse` a `JSONResponse`? Afterall, it's just the response itself which happens to be JSON. Perhaps `HTTPResponse` could be `RAWResponse` and uses for non-JSON payloads - perhaps with a content-type? More focus on the payload than how it was retrieved.

**Comment by samcoe** [comment]
Closing as we decided against exporting `httpmock` in favor of migrating to `gock` in https://github.com/cli/go-gh/pull/31.

**Pull request closed without merging**